### DISCUSSION TITLE
Use a Given PyTorch Data Loader

### DISCUSSION THREAD

**Discussion by RoyiAvital** [skorch-dev/skorch#980] [Q&A] [answered]
Is there a way to use a given PyTorch data loader instead of just a data set?

**Comment by BenjaminBossan** [comment, marked as answer]
You cannot pass a `DataLoader` instance to `fit`.

You can change the _class_ of the `DataLoader` by passing it to the `iterator_train` and `iterator_valid` argument.

If you want to have control over the actual `DataLoader` instance, you'd have to subclass the net and override the [`get_iterator`](https://github.com/skorch-dev/skorch/blob/5b222a5897071c35a3ab0e0663796be366fdc954/skorch/net.py#L1756) method. You can check the code, it should be quite straightforward to make it return something custom built.

**Reply by RoyiAvital** [reply]
I see.

So if I have `dsTrain` and `dsValid`, each is a data set (PyTorch `DataSet`) I can do something like:

```python
netCls = NeuralNetClassifier(
    modelCls,
    train_split = predefined_split(dsValid),
)

netCls.train_split.stratified = False

netCls = netCls.fit(dsTrain, None)
```

Then all data in `dsTrain` will be used for training and all data in `dsTValid` will be used for validation, right?
Or should I set `train_split` to `None` as well?

By the way, any way to set `netCls.train_split.stratified = False` in the `NeuralNetClassifier()` construction?
I tried `train_split__stratified = False`, yet it didn't work.

**Reply by ottonemo** [reply]
> Then all data in `dsTrain` will be used for training and all data in `dsValid` will be used for validation, right?

Correct!

>  Or should I set `train_split` to None as well?

That would deactivate the validation split completely (no validation run/metrics).

> By the way, any way to set `netCls.train_split.stratified = False` in the `NeuralNetClassifier()` construction?

In a predefined split there is no stratification (splitting is already done). 
The default value for the `train_split` parameter is `skorch.dataset.ValidSplit(5)` which, depending on the type of network you use, has stratification enabled or disabled (classification uses stratification, regression doesn't).

So, in essence, if you use the `ValidSplit` class for the `train_split` parameter, then you can deactivate stratification by passing `stratified=False` to `ValidSplit` but if you use a predefined split, there's nothing to deactivate because *you* already did the split.

**Reply by RoyiAvital** [reply]
@ottonemo , I see. So in general (Default split), to have the effect of `netCls.train_split.stratified = False` in the construction of the classifier I can do something like:

```python
netCls = NeuralNetClassifier(
    modelCls,
    ValidSplit__stratified = False,
)
```

**Reply by githubnemo** [reply]
Almost :)

```python
netCls = NeuralNetClassifier(
    modelCls,
    train_split__stratified=False,
)
```

**Reply by RoyiAvital** [reply]
Strange, I think I tried it:

> I tried `train_split__stratified = False`, yet it didn't work.

But maybe I tried it in the case you wrote above, with the pre defined split.

**Reply by githubnemo** [reply]
Sorry, you are absolutely right. I thought we passed these arguments to the split as well but I remembered wrongly. It also doesn't really make sense since the `VaildSplit` is passed as an instance and not as a class (in contrast to, e .g., the module), so we can directly set the parameters we want.

```python
from skorch.dataset import ValidSplit

netCls = NeuralNetClassifier(
    modelCls,
    train_split=ValidSplit(5, stratified=False),
)
```

**Comment by ottonemo** [comment]
To add to @BenjaminBossan's answer:

> If you want to have control over the actual DataLoader instance, you'd have to subclass the net and override the [get_iterator](https://github.com/skorch-dev/skorch/blob/5b222a5897071c35a3ab0e0663796be366fdc954/skorch/net.py#L1756) method. You can check the code, it should be quite straightforward to make it return something custom built.

While true you can also pass a generator function to the net instance that yields from an existing dataloader or constructs one on the fly.
For example:

```python
def train_iterator(dataset, **kwargs):
    loader = DataLoader(dataset, **kwargs)

    for X, y in loader:
        yield X, y

net = NeuralNet(
    ...,
    iterator_train=train_iterator,
    ....,
)
```
